packages/material-ui-utils/src/index.ts

```diff
@@ -8,3 +8,21 @@ export { default as getDisplayName } from './getDisplayName';
 export { default as HTMLElementType } from './HTMLElementType';
 export { default as ponyfillGlobal } from './ponyfillGlobal';
 export { default as refType } from './refType';
+export { default as capitalize } from './capitalize';
+export { default as createChainedFunction } from './createChainedFunction';
+export { default as debounce } from './debounce';
+export { default as deprecatedPropType } from './deprecatedPropType';
+export { default as isMuiElement } from './isMuiElement';
+export { default as ownerDocument } from './ownerDocument';
+export { default as ownerWindow } from './ownerWindow';
+export { default as requirePropFactory } from './requirePropFactory';
+export { default as setRef } from './setRef';
+export { default as unstable_useEnhancedEffect } from './useEnhancedEffect';
+export { default as unstable_useId } from './useId';
```

Comment:
Do we need the unstable prefix if we treat all the modules here private?

Reply:
If they're here, they're public.

Reply:
I forgot where I said but I would be way more comfortable by renaming `/utils` to `internal` or even `unsafe_internal`. There are already plenty of packages depending on `/types` and I suspect the same happened with `/utils`. 

Or we release it as an alpha. 

Either way, I agree that the current state of /utils being on the same release line as /core (5.x) does not signal that the package is private. It's simply not enough that we treat it as private since we have some responsibility to not clutter npm with more underdocumented packages.

Reply:
Should we then proceed with https://github.com/mui-org/material-ui/pull/23270 move all utils in `unstyled` and export them as `unstable_`? Then for avoiding breaking changes, we can reuse them in the core and just reexport them.

Reply:
> Then for avoiding breaking changes, we can reuse them in the core and just reexport them.

I don't understand what that solves. 

If we change the API of `useId` and only we consume `useId` then everything is fine. 
If `useId` can also be used by library consumers directly then we can only release a breaking change either by 
1. bumping the major version number of the package
1. bumping the pre-release number if it's an alpha
1. do nothing (but documentation) if it's prefixed as `unstable_`
   `unstable_` is still only an informal convention so we should still be careful.

Reply:
Sorry my question was misleading. We already have the useId as `unstable_useId` exported from `@material-ui/core/utils`, so I was thinking instead of duplicating the code we can re-export this utility again from there. 

My question was more of: Should we move all these utils to `@material-ui/utils` as this PR is doing and export them all as `unstable_` or should we move them to the new`@material-ui/unstyled` package and again to export them as `unstable_`. I don't have strong preference, but we need to keep them either in `utils` or `unstyled`, so that those can be reused from the `@material-ui/unstyled` package too. Based on the discussion seems like they will be anyway exported as `unstable_`, so I don't see why one package would be prefered than the other. I guess we just need to decide where to export them from.

Reply:
This issue #20353 is relevant to the problem. The utils exported from `@material-ui/core` are public.

Reply:
> or should we move them to the new@material-ui/unstyled package and again to export them as unstable_

Considering that some of the utils are required by the system, it would mean the system needs to depend on unstyled. I assume that it's not something we want because the two are solving independent problems.

> Should we move all these utils to @material-ui/utils as this PR is doing and export them all as unstable_

I like this option +1

Reply:
Alright, let me update the exports then

Reply:
Done, everything from utils that comes from `@material-ui/core/utils` is exported as `unstable_`

Reply:
We're importing `unstable_createFunction` from /utils and re-export it as `createChainedFunction` from /core/utils. I don't know how this is supposed to work: If /utils can change the API of createChainedFunction on every release and /core pins /utils to major then createChainedFunction from /core/utils can also change on every release and is therefore unstable.

What did we try to accomplish by **adding** (and not moving) /core/utils to /utils?

Reply:
I don't want to make changes to the existing `core/utils` as we have imports all over the project from there. I started like that in https://github.com/mui-org/material-ui/pull/23203 prepared codemods as well, but we decided to go wtih the minimal changes necessary for unblocking ourselves, which is to move the utils to core, but still re-export them from the core in order to avoid changes in the core.